node/tests/node.rs

```diff
@@ -213,3 +199,521 @@ async fn peer_discovery() {
+
+    // request entire store range
+    let received_all_headers = client
+        .p2p()
+        .get_verified_headers_range(&received_genesis, 19)
+        .await
+        .unwrap();
+    assert_eq!(server_headers[1..], received_all_headers);
+
+    // reqest more headers than available in store
+    let out_of_bounds = client
+        .p2p()
+        .get_verified_headers_range(&received_genesis, 20)
+        .await
+        .unwrap_err();
+    assert!(matches!(
+        out_of_bounds,
+        P2pError::Exchange(ExchangeError::InvalidResponse)
+    ));
+
+    // request unknown hash
+    let unstored_header = header_generator.next_of(&server_headers[0]);
+    let unexpected_hash = client
+        .p2p()
+        .get_header(unstored_header.hash())
+        .await
+        .unwrap_err();
+    assert!(matches!(
+        unexpected_hash,
+        P2pError::Exchange(ExchangeError::HeaderNotFound)
+    ));
+
+    // request unknown height
+    let unexpected_height = client.p2p().get_header_by_height(21).await.unwrap_err();
+    assert!(matches!(
+        unexpected_height,
+        P2pError::Exchange(ExchangeError::HeaderNotFound)
+    ));
+}
+
+#[tokio::test]
+async fn client_server_invalid_requests_test() {
+    // Server Node
+    let server = Node::new(NodeConfig {
+        store: gen_filled_store(20).0,
+        ..listening_test_node_config()
+    })
+    .await
+    .unwrap();
+
+    // give server a sec to breathe, otherwise occiasionally client has problems with connecting
+    sleep(Duration::from_millis(100)).await;
+    let server_addrs = server.p2p().listeners().await.unwrap();
+
+    // Client node
+    let client = Node::new(NodeConfig {
+        p2p_bootstrap_peers: server_addrs.clone(),
+        ..test_node_config()
+    })
+    .await
+    .unwrap();
+
+    client.p2p().wait_connected().await.unwrap();
+
+    let none_data = client
+        .p2p()
+        .exchange_header_request(HeaderRequest {
+            data: None,
+            amount: 1,
+        })
+        .await
+        .unwrap_err();
+    assert!(matches!(
+        none_data,
+        P2pError::Exchange(ExchangeError::InvalidRequest)
+    ));
+
+    let zero_amount = client
+        .p2p()
+        .exchange_header_request(HeaderRequest {
+            data: Some(header_request::Data::Origin(5)),
+            amount: 0,
+        })
+        .await
+        .unwrap_err();
+    assert!(matches!(
+        zero_amount,
+        P2pError::Exchange(ExchangeError::InvalidRequest)
+    ));
+
+    let malformed_hash = client
+        .p2p()
+        .exchange_header_request(HeaderRequest {
+            data: Some(header_request::Data::Hash(vec![0; 31])),
+            amount: 1,
+        })
+        .await
+        .unwrap_err();
+    assert!(matches!(
+        malformed_hash,
+        P2pError::Exchange(ExchangeError::InvalidRequest)
+    ));
+}
+
+#[tokio::test]
+async fn head_selection_with_multiple_peers() {
```

Comment:
I can not see why this will fail with the Syncer. Can you explain?

Reply:
when you connect the node that have the height 15, it will announce it on the header-sub and syncer will try to find it by requesting next headers from the peers it knows

Reply:
hmm, unless it doesn't announce it on header-sub :thinking:

Reply:
The flow that I'm implementing right now is like this:

* `Syncer` get HEAD via the Exchange protocol.
* `Syncer` initialize the HeaderSub with the HEAD that it got
* HeaderSub now can inform `Syncer` for any new headers via a watcher
* HeaderSub only propagates messages, it doesn't publish on its own.

Reply:
per offline conversation, our test network shouldn't publish headersub messages by itself, so stores will remain static.